motionplan/ptgIK.go

```diff
@@ -0,0 +1,141 @@
+)
+
+// ptgGridSim will take a PrecomputePTG, and simulate out a number of trajectories through some requested time/distance for speed of lookup
+// later. It will store the trajectories in a grid data structure allowing relatively fast lookups.
+type ptgIK struct {
+	refDist         float64
+	simPTG          tpspace.PrecomputePTG
+	ptgFrame        referenceframe.Frame
+	fastGradDescent *NloptIK
+
+	gridSim tpspace.PTG
+
+	mu        sync.RWMutex
+	trajCache map[float64][]*tpspace.TrajNode
+}
+
+// NewPTGIK creates a new PTG by simulating a PrecomputePTG for some distance, then cacheing the results in a grid for fast lookup.
+func NewPTGIK(simPTG tpspace.PrecomputePTG, logger golog.Logger, refDist float64, randSeed int) (tpspace.PTG, error) {
+	ptgFrame, err := tpspace.NewPTGIKFrame(simPTG, refDist)
+	if err != nil {
+		return nil, err
+	}
+
+	nlopt, err := CreateNloptIKSolver(ptgFrame, logger, 1, defaultEpsilon*defaultEpsilon, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// create an ends-only grid sim for quick end-of-trajectory calculations
+	gridSim, err := tpspace.NewPTGGridSim(simPTG, 0, refDist, true)
+	if err != nil {
+		return nil, err
+	}
+
+	ptg := &ptgIK{
+		refDist:         refDist,
+		simPTG:          simPTG,
+		ptgFrame:        ptgFrame,
+		fastGradDescent: nlopt,
+		gridSim:         gridSim,
+		trajCache:       map[float64][]*tpspace.TrajNode{},
+	}
+
+	return ptg, nil
+}
+
+func (ptg *ptgIK) CToTP(ctx context.Context, distFunc func(spatialmath.Pose) float64) (*tpspace.TrajNode, error) {
+	solutionGen := make(chan *IKSolution, 1)
+	seedInput := []referenceframe.Input{{math.Pi / 2}, {ptg.refDist / 2}} // random value to seed the IK solver
+	goalMetric := func(state *State) float64 {
+		return distFunc(state.Position)
+	}
+	// Spawn the IK solver to generate a solution
+	err := ptg.fastGradDescent.Solve(ctx, solutionGen, seedInput, goalMetric, nloptSeed)
+	// We should have zero or one solutions
+	var solved *IKSolution
+	select {
+	case solved = <-solutionGen:
+	default:
+	}
+	close(solutionGen)
+	if err != nil || solved == nil || solved.Configuration[1].Value < defaultZeroDist {
+		return ptg.gridSim.CToTP(ctx, distFunc)
+	}
+
+	if solved.Partial {
+		gridNode, err := ptg.gridSim.CToTP(ctx, distFunc)
+		if err == nil {
+			// Check if the grid has a better solution
+			if distFunc(gridNode.Pose) < solved.Score {
+				return gridNode, nil
+			}
+		}
+	}
+
+	traj, err := ptg.Trajectory(solved.Configuration[0].Value, solved.Configuration[1].Value)
+	if err != nil {
+		return nil, err
+	}
+	return traj[len(traj)-1], nil
+}
+
+func (ptg *ptgIK) RefDistance() float64 {
+	return ptg.refDist
+}
+
+func (ptg *ptgIK) Trajectory(alpha, dist float64) ([]*tpspace.TrajNode, error) {
+	ptg.mu.RLock()
+	precomp := ptg.trajCache[alpha]
+	ptg.mu.RUnlock()
+	if precomp != nil {
+		if precomp[len(precomp)-1].Dist >= dist {
+			// Cacheing here provides a ~33% speedup to a solve call
```

Comment:
Can you explain why this is true? Are we likely to call `Trajectory()` with the same `dist` value again? Do we call `Trajectory` with discrete values of `dist`?

Reply:
First, we currently call `Trajectory()` twice per node that we add; once when we determine its viability and a second time when we actually add it (and any intermediate nodes, optionally) to the tree. This of course speeds that up, though if we didn't have this caching in place, we could save that trajectory in the `candidate` data structure as well.

But secondly, and importantly, the algorithm spends most of its iterations extending the map by trajectories of length `refDist` as the two maps come together; for this we use a lookup of a discreized set of alpha values already computed at refDist, rather than running IK. These tend to be repetitive, as a straight line is a straight line.

Reply:
Thank you!